File: src/usecases/SearchExerciseVideos.ts
import z from "zod";

import { ExternalServiceError } from "../errors/index.js";
import { env } from "../lib/env.js";

const YOUTUBE_SEARCH_URL = "https://www.googleapis.com/youtube/v3/search";
const MAX_VIDEOS = 3;

const YouTubeSearchResponseSchema = z.object({
  items: z.array(
    z.object({
      id: z.object({ videoId: z.string().min(1) }),
      snippet: z.object({
        title: z.string(),
        channelTitle: z.string(),
        thumbnails: z.object({
          high: z.object({ url: z.url() }).optional(),
          medium: z.object({ url: z.url() }).optional(),
          default: z.object({ url: z.url() }).optional(),
        }),
      }),
    })
  ),
});

interface InputDto {
  exerciseName: string;
}

interface OutputDto {
  videos: Array<{
    title: string;
    channelTitle: string;
    url: string;
    thumbnailUrl: string;
  }>;
  searchUrl: string;
}

const buildSearchQuery = (exerciseName: string): string =>
  `${exerciseName.trim()} execução correta`;

export const buildExerciseVideoSearchUrl = (exerciseName: string): string => {
  const url = new URL("https://www.youtube.com/results");
  url.searchParams.set("search_query", buildSearchQuery(exerciseName));
  return url.toString();
};

export class SearchExerciseVideos {
  async execute(dto: InputDto): Promise<OutputDto> {
    const searchUrl = buildExerciseVideoSearchUrl(dto.exerciseName);

    if (!env.YOUTUBE_API_KEY) {
      return { videos: [], searchUrl };
    }

    const url = new URL(YOUTUBE_SEARCH_URL);
    url.searchParams.set("part", "snippet");
    url.searchParams.set("type", "video");
    url.searchParams.set("maxResults", String(MAX_VIDEOS));
    url.searchParams.set("relevanceLanguage", "pt");
    url.searchParams.set("safeSearch", "strict");
    url.searchParams.set("q", buildSearchQuery(dto.exerciseName));
    url.searchParams.set("key", env.YOUTUBE_API_KEY);

    const response = await fetch(url);
    if (!response.ok) {
      throw new ExternalServiceError(
        `YouTube search failed with status ${response.status}`
      );
    }

    const parsed = YouTubeSearchResponseSchema.safeParse(await response.json());
    if (!parsed.success) {
      throw new ExternalServiceError("Unexpected YouTube search response");
    }

    return {
      videos: parsed.data.items.map((item) => {
        const { thumbnails } = item.snippet;
        return {
          title: item.snippet.title,
          channelTitle: item.snippet.channelTitle,
          url: `https://www.youtube.com/watch?v=${item.id.videoId}`,
          thumbnailUrl:
            thumbnails.high?.url ??
            thumbnails.medium?.url ??
            thumbnails.default?.url ??
            `https://i.ytimg.com/vi/${item.id.videoId}/hqdefault.jpg`,
        };
      }),
      searchUrl,
    };
  }
}

File: backend/FitAi.Api/UseCases/Coach/SearchExerciseVideos.cs
using System.Text.Json;
using FitAi.Api.Errors;
using FitAi.Api.Options;
using FitAi.Contracts;
using Microsoft.Extensions.Options;

namespace FitAi.Api.UseCases.Coach;

/// <summary>Busca vídeos de execução no YouTube. Sem chave configurada, devolve só o link de busca.</summary>
public sealed class SearchExerciseVideos(HttpClient http, IOptions<YouTubeOptions> options)
{
    private const int MaxVideos = 3;

    public sealed record Input(string ExerciseName);

    public sealed record Output(IReadOnlyList<ExerciseVideoResponse> Videos, string SearchUrl);

    public static string BuildSearchQuery(string exerciseName) => $"{exerciseName.Trim()} execução correta";

    public static string BuildSearchUrl(string exerciseName) =>
        "https://www.youtube.com/results?search_query=" + Uri.EscapeDataString(BuildSearchQuery(exerciseName));

    public async Task<Output> ExecuteAsync(Input input, CancellationToken ct = default)
    {
        var searchUrl = BuildSearchUrl(input.ExerciseName);
        var apiKey = options.Value.ApiKey;
        if (string.IsNullOrWhiteSpace(apiKey)) return new Output([], searchUrl);

        var url = "https://www.googleapis.com/youtube/v3/search?part=snippet&type=video"
            + $"&maxResults={MaxVideos}&relevanceLanguage=pt&safeSearch=strict"
            + $"&q={Uri.EscapeDataString(BuildSearchQuery(input.ExerciseName))}&key={Uri.EscapeDataString(apiKey)}";

        using var response = await http.GetAsync(url, ct);
        if (!response.IsSuccessStatusCode)
        {
            throw new ExternalServiceException($"YouTube search failed with status {(int)response.StatusCode}");
        }

        try
        {
            using var json = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync(ct), cancellationToken: ct);
            var videos = new List<ExerciseVideoResponse>();
            foreach (var item in json.RootElement.GetProperty("items").EnumerateArray())
            {
                var videoId = item.GetProperty("id").GetProperty("videoId").GetString();
                if (string.IsNullOrEmpty(videoId)) continue;
                var snippet = item.GetProperty("snippet");
                var thumbnails = snippet.GetProperty("thumbnails");
                string? thumbnail = null;
                foreach (var size in new[] { "high", "medium", "default" })
                {
                    if (thumbnails.TryGetProperty(size, out var t)) { thumbnail = t.GetProperty("url").GetString(); break; }
                }
                videos.Add(new ExerciseVideoResponse(
                    snippet.GetProperty("title").GetString() ?? "",
                    snippet.GetProperty("channelTitle").GetString() ?? "",
                    $"https://www.youtube.com/watch?v={videoId}",
                    thumbnail ?? $"https://i.ytimg.com/vi/{videoId}/hqdefault.jpg"));
            }
            return new Output(videos, searchUrl);
        }
        catch (Exception e) when (e is JsonException or KeyNotFoundException or InvalidOperationException)
        {
            throw new ExternalServiceException("Unexpected YouTube search response");
        }
    }
}
